Kianoosh76/IBS
# [ibs-90] Make the Gitlab connection polymorphic

## Description

- This PR introduces several changes aimed at enhancing the functionality of the version control system (VCS) integration within the project. The key changes include:
  1. Introduction of an abstract VCS class (`AbstractVCS`) along with concrete implementations for GitLab (`GitLabVCS`) and a mock VCS (`MockVCS`).
  2. Addition of a `user_id` column to the `User` model to facilitate user-related operations in the VCS.
  3. Mocking of certain functionalities in the `MockVCS` class to simulate GitLab API responses and facilitate testing.

## Changes Made

- This PR addresses the following tasks:
  1. Implement AbstractVCS and concrete VCS classes.
  2. Add user_id column to the User model.
  3. Implement VCS-related methods.
  4. Mock certain functionalities in the MockVCS class.

## Test Guidelines

1. Read through and review the newly added VCS files
2. Run basic tests such as `npm test registration`, `npm test login`, `npm test add_course`, `npm test get_course` etc to verify that the current login flow still works with the User model update.

## Checklist

### For PR author:
- [ ] I have added appropriate comments to the code.
- [ ] I have ensured code style and formatting are consistent.
- [ ] I have included unit tests or integration tests where applicable.
- [ ] I have considered edge cases and handled them appropriately.
- [ ] I have celebrated my accomplishments 🎉

### For PR reviewer:
- [ ] I have reviewed the code changes.
- [ ] I have verified that comments are clear and helpful.
- [ ] I have ensured code style and formatting align with project standards.
- [ ] I have examined any included tests and their coverage.
- [ ] I have considered potential edge cases and provided feedback on them.
